    The objective is to find the association rules to help supermarket owners find new marketing plans to improve their sales.

In [1]:
toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],

    ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],

    ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],

    ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],

    ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

Run the apriori algorithm on the provided toy_dataset. Interpret the results.

In [2]:
import mlxtend
import pandas as pd 
from mlxtend.preprocessing import TransactionEncoder  # Using `TransactionEncoder`
TE = TransactionEncoder()
TE_ary= TE.fit_transform(toy_dataset)  #Apply one-hot-encoding on our toy_dataset
TE_ary

array([[False, False,  True, False, False,  True, False,  True, False,
         True,  True, False],
       [False, False,  True, False, False,  True, False, False, False,
         True,  True,  True],
       [ True, False, False,  True, False,  True,  True,  True, False,
        False, False, False],
       [ True,  True,  True,  True,  True, False, False,  True, False,
        False, False, False],
       [False,  True, False, False, False,  True, False,  True,  True,
         True, False, False]])

In [3]:
toy_df= pd.DataFrame(TE_ary, columns= TE.columns_) #Creating a new DataFrame from our Numpy array
toy_df

,Dress,Earrings,Hat,Heels,Necklace,Pants,Sandals,Scarf,Shirt,Skirt,Sneakers,Sunglasses
0,False,False,True,False,False,True,False,True,False,True,True,False
1,False,False,True,False,False,True,False,False,False,True,True,True
2,True,False,False,True,False,True,True,True,False,False,False,False
3,True,True,True,True,True,False,False,True,False,False,False,False
4,False,True,False,False,False,True,False,True,True,True,False,False


Now that the dataset has been transformed using one-hot encoding, we can move on to **extracting frequent itemsets** using the **Apriori algorithm**.

In [4]:
from mlxtend.frequent_patterns import apriori
apriori(toy_df, min_support=0.6) # Support code 

,support,itemsets
0,0.6,(2)
1,0.8,(5)
2,0.8,(7)
3,0.6,(9)
4,0.6,"(5, 7)"
5,0.6,"(9, 5)"


In [5]:
frequent_itemsets=apriori(toy_df, min_support=0.6, use_colnames=True)  #Instead of column indices we can use column names
frequent_itemsets #Support with column names

,support,itemsets
0,0.6,(Hat)
1,0.8,(Pants)
2,0.8,(Scarf)
3,0.6,(Skirt)
4,0.6,"(Pants, Scarf)"
5,0.6,"(Pants, Skirt)"


Once we have our frequent itemsets, we can extract **association rules** using other interesting metrics like **confidence** and **lift**.

In [6]:
from mlxtend.frequent_patterns import association_rules
association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7) # associate items

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(Pants),(Scarf),0.8,0.8,0.6,0.75,0.9375,1.0,-0.04,0.8,-0.25,0.60,-0.250,0.750
1,(Scarf),(Pants),0.8,0.8,0.6,0.75,0.9375,1.0,-0.04,0.8,-0.25,0.60,-0.250,0.750
2,(Pants),(Skirt),0.8,0.6,0.6,0.75,1.2500,1.0,0.12,1.6,1.00,0.75,0.375,0.875
3,(Skirt),(Pants),0.6,0.8,0.6,1.00,1.2500,1.0,0.12,inf,0.50,0.75,1.000,0.875


In addition to **confidence**, we can also use **lift** as a metric to extract more meaningful association rules.

In [7]:
from mlxtend.frequent_patterns import association_rules
association_rules(frequent_itemsets,metric="lift",min_threshold=1.25)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(Skirt),(Pants),0.6,0.8,0.6,1.0,1.25,1.0,0.12,inf,0.5,0.75,1.0,0.875


    Top Association Rules Interpretation
**1. (Skirt) → (Pants)**

- Support: 0.6, This pair appears in 60% of all transactions.

- Confidence: 1.0, Every time Skirt appears, Pants also appears.

- Lift: 1.25, lift >1, so the likelihood of Pants being bought increases by 25% when Skirt is present.

- Certainty: 1.0, Strongest possible certainty that Pants follows Skirt.

This is a very strong rule. If a customer buys a Skirt, they always buy Pants too. Great for bundling or recommendation systems.

**2. (Pants) → (Skirt)**

- Support: 0.6

- Confidence: 0.75, 75% of Pants purchases also include Skirt.

- Lift: 1.25, Still a positive association.

- Certainty: 0.375, Lower than the reverse rule.

While Pants often lead to Skirt, it's not guaranteed. Still a good cross-sell opportunity, but not as strong as the reverse.

**3. (Scarf) → (Pants) and (Pants) → (Scarf)**

- Support: 0.6

- Confidence: 0.75 in both directions

- Lift: 0.9375, Slightly below 1, which suggests a negative or neutral association

- Certainty: -0.25 , Weak and potentially misleading rule.

Despite high confidence, the lift < 1 and negative leverage suggest this rule is not truly useful. These items co-occur often, but not more than random chance would predict.

*******************************************************************************************************************************

In [8]:
data = pd.read_csv("C:/Users/ABL/Documents/Codes/checkp_python/ML/Association_rules_checkp/Customer_purchases_history/DATASET/Market_Basket_Optimisation.csv", header=None)
data.head()

,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19
0,shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
1,burgers,meatballs,eggs,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,chutney,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,turkey,avocado,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,mineral water,milk,energy bar,whole wheat rice,green tea,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
import csv # Imports the csv module to handle CSV file operations

# Defines a function to read a CSV file and return its contents as a list of lists
def read_csv_as_list_of_lists(filename):
    data = []  # Initializes an empty list to store rows from the CSV
    with open(filename, newline='', encoding='utf-8') as csvfile: # Opens the CSV file in read mode
        reader = csv.reader(csvfile)  # Creates a CSV reader object
        for row in reader:
            # Checks if the row is not empty before appending
            if row:
                data.append(row) # Adds the non-empty row to the data list
    return data  # Returns the complete list of rows

# Example usage
filename = 'C:\\Users\\ABL\\Documents\\Codes\\checkp_python\\ML\\Association_rules_checkp\\Customer_purchases_history\\DATASET\\Market_Basket_Optimisation.csv'  # Replace with your actual file path
rows = read_csv_as_list_of_lists(filename)  # Calls the function and stores the result in 'rows'
rows # Displays the loaded data (each row is a list of items purchased by a customer)

In [10]:
# Flatten list of lists into one pandas Series
df = pd.Series([item for sublist in rows for item in sublist], name="Item")

# Count frequencies
freq = df.value_counts().reset_index()
freq.columns = ['Items', 'Count']

In [11]:
import plotly.express as px # Creates a bar chart from the DataFrame 'freq'
fig = px.bar(freq, x='Items', # DataFrame containing item frequencies
             y='Count',
             title='Frequency of Items')
fig.show()

In [12]:
# Select the top 10 most frequent items
top10 = freq.head(10)
# Create a bar chart
fig = px.bar(top10, # DataFrame with top 10 items
             x='Items', 
             y='Count', 
             title='Top 10 Most Frequent Items in The Dataset',
             color='Count', # Bar color intensity based on count
             color_continuous_scale='Blues')

# Sort x-axis categories by total count in descending order
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

In [13]:
# Initialize an empty list to store item pairs
pairs = []

# Loop through each transaction (basket) in the dataset
for transaction in rows:
    # Remove duplicate items within the same basket
    unique_items = list(set(transaction))
    
    # Generate all unique pairs of items from the basket
    for i in range(len(unique_items)):
        for j in range(i + 1, len(unique_items)):
            # Sort the pair alphabetically to avoid duplicates like ['A', 'B'] and ['B', 'A']
            pair = sorted([unique_items[i], unique_items[j]])
            # Add the pair to the list
            pairs.append(pair)

# Convert the list of pairs into a DataFrame with two columns: Item1 and Item2
df_pairs = pd.DataFrame(pairs, columns=['Item1', 'Item2'])

# Count how many times each unique pair appears in the dataset
pair_counts = df_pairs.value_counts().reset_index(name='Count')

# Select the top 10 most frequent item pairs
top_pairs = pair_counts.head(10)

# Create a horizontal bar chart using Plotly
fig = px.bar(
    top_pairs,  # DataFrame containing top item pairs and their counts
    x='Count',  # Frequency of each pair on the x-axis
    y=top_pairs['Item1'] + ' + ' + top_pairs['Item2'],  # Concatenate item names for y-axis labels
    orientation='h',  # Horizontal bars
    text='Count',  # Show count values on each bar
    title='Top 10 Most Frequently Bought-Together Pairs (No itertools)',  # Chart title
    color='Count',  # Color bars based on frequency
    color_continuous_scale='Blues'  # Use a blue gradient for visual appeal
)

# Position the count labels outside the bars
fig.update_traces(textposition='outside')

# Customize axis titles
fig.update_layout(yaxis_title='Item Pair', xaxis_title='Frequency')

# Display the chart
fig.show()


In [14]:
TE = TransactionEncoder()
TE_ary= TE.fit_transform(rows)  #Apply one-hot-encoding on our toy_dataset
TE_ary

array([[False,  True,  True, ...,  True, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False]])

In [15]:
df= pd.DataFrame(TE_ary, columns= TE.columns_) #Creating a new DataFrame from our Numpy array
df

,asparagus,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,black tea,blueberries,...,turkey,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,True,True,False,True,False,False,False,False,False,...,False,True,False,False,True,False,False,True,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,True,False,False,False,False,False,...,True,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7496,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7497,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7498,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7499,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


In [16]:
from mlxtend.frequent_patterns import apriori
apriori(df, min_support=0.05) # Support code 

,support,itemsets
0,0.087188,(15)
1,0.081056,(17)
2,0.059992,(23)
3,0.163845,(25)
4,0.080389,(30)
5,0.051060,(31)
6,0.179709,(37)
7,0.079323,(40)
8,0.170911,(43)
9,0.063325,(48)


In [17]:
frequent_itemsets=apriori(df,min_support=0.05, use_colnames=True)  #Instead of column indices we can use column names
frequent_itemsets #Support with column names

,support,itemsets
0,0.087188,(burgers)
1,0.081056,(cake)
2,0.059992,(chicken)
3,0.163845,(chocolate)
4,0.080389,(cookies)
5,0.051060,(cooking oil)
6,0.179709,(eggs)
7,0.079323,(escalope)
8,0.170911,(french fries)
9,0.063325,(frozen smoothie)


In [18]:
from mlxtend.frequent_patterns import association_rules
association_rules(frequent_itemsets,metric="confidence",min_threshold=0.01) # associate items

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(mineral water),(chocolate),0.238368,0.163845,0.052660,0.220917,1.348332,1.0,0.013604,1.073256,0.339197,0.150648,0.068256,0.271158
1,(chocolate),(mineral water),0.163845,0.238368,0.052660,0.321400,1.348332,1.0,0.013604,1.122357,0.308965,0.150648,0.109018,0.271158
2,(mineral water),(eggs),0.238368,0.179709,0.050927,0.213647,1.188845,1.0,0.008090,1.043158,0.208562,0.138707,0.041372,0.248515
3,(eggs),(mineral water),0.179709,0.238368,0.050927,0.283383,1.188845,1.0,0.008090,1.062815,0.193648,0.138707,0.059103,0.248515
4,(mineral water),(spaghetti),0.238368,0.174110,0.059725,0.250559,1.439085,1.0,0.018223,1.102008,0.400606,0.169312,0.092566,0.296796
5,(spaghetti),(mineral water),0.174110,0.238368,0.059725,0.343032,1.439085,1.0,0.018223,1.159314,0.369437,0.169312,0.137421,0.296796


In [19]:
from mlxtend.frequent_patterns import association_rules
association_rules(frequent_itemsets,metric="lift",min_threshold=1.25)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(mineral water),(chocolate),0.238368,0.163845,0.052660,0.220917,1.348332,1.0,0.013604,1.073256,0.339197,0.150648,0.068256,0.271158
1,(chocolate),(mineral water),0.163845,0.238368,0.052660,0.321400,1.348332,1.0,0.013604,1.122357,0.308965,0.150648,0.109018,0.271158
2,(mineral water),(spaghetti),0.238368,0.174110,0.059725,0.250559,1.439085,1.0,0.018223,1.102008,0.400606,0.169312,0.092566,0.296796
3,(spaghetti),(mineral water),0.174110,0.238368,0.059725,0.343032,1.439085,1.0,0.018223,1.159314,0.369437,0.169312,0.137421,0.296796


**Low Support Thresholds Work**

Support = 0.05 means we're looking for itemsets that appear in 5% of transactions.

The dataset is diverse, customers buy very different combinations of items.

**No Strong Rules for "mineral water"**

we discovered that mineral water appears in 23.8% of all transactions.

"mineral water" may be popular, but not predictably paired with other items.



**chocolate → mineral water**

Support: 5.3% → This pair appears in 5.3% of all transactions.

Confidence: 32.1% → When chocolate is bought, mineral water is also bought 32.1% of the time.

Lift: 1.35 → This is 35% more likely than random chance — a meaningful association.


**spaghetti → mineral water**

Support: 5.9%

Confidence: 34.3%

Lift: 1.44 

spaghetti buyers are 44% more likely to also buy mineral water.